core/src/main/java/bisq/core/trade/protocol/bsqswap/tasks/buyer/BuyerFinalizeTx.java

```diff
@@ -0,0 +1,151 @@
+import org.bitcoinj.core.TransactionInput;
+import org.bitcoinj.core.TransactionOutPoint;
+import org.bitcoinj.core.TransactionOutput;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import lombok.extern.slf4j.Slf4j;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+@Slf4j
+public class BuyerFinalizeTx extends BsqSwapTask {
+    @SuppressWarnings({"unused"})
+    public BuyerFinalizeTx(TaskRunner<BsqSwapTrade> taskHandler, BsqSwapTrade bsqSwapTrade) {
+        super(taskHandler, bsqSwapTrade);
+    }
+
+    @Override
+    protected void run() {
+        try {
+            runInterceptHook();
+
+            BsqSwapTradePeer tradePeer = bsqSwapProtocolModel.getTradePeer();
+            TradeWalletService tradeWalletService = bsqSwapProtocolModel.getTradeWalletService();
+            BtcWalletService btcWalletService = bsqSwapProtocolModel.getBtcWalletService();
+            NetworkParameters params = btcWalletService.getParams();
+
+            Transaction sellersTransaction = btcWalletService.getTxFromSerializedTx(tradePeer.getTx());
+
+            // The inputs from the deserialized tx do not have connected outpoints and no value assigned.
+            // We reapply that by using the RawTransactionInputs
+            int buyersInputSize = Objects.requireNonNull(bsqSwapProtocolModel.getInputs()).size();
+            List<TransactionInput> sellersInputs = sellersTransaction.getInputs().stream()
+                    .filter(i -> i.getIndex() >= buyersInputSize)
+                    .collect(Collectors.toList());
+
+            List<RawTransactionInput> rawInputs = Objects.requireNonNull(tradePeer.getInputs());
+            List<TransactionInput> sellersBtcInputs = new ArrayList<>();
+            for (int i = 0; i < sellersInputs.size(); i++) {
+                TransactionInput transactionInput = sellersInputs.get(i);
+                RawTransactionInput rawTransactionInput = rawInputs.get(i);
+                Coin value = Coin.valueOf(rawTransactionInput.value);
+                Transaction parentTx = new Transaction(params, rawTransactionInput.parentTransaction);
+                TransactionOutPoint transactionOutPoint = new TransactionOutPoint(params, rawTransactionInput.index, parentTx);
+                TransactionInput result = new TransactionInput(params,
+                        parentTx,
+                        transactionInput.getScriptBytes(),
+                        transactionOutPoint,
+                        value);
+                result.setWitness(transactionInput.getWitness());
+                sellersBtcInputs.add(result);
+            }
+
+            TransactionOutput sellersBsqPayout = sellersTransaction.getOutput(0);
+            tradePeer.setBsqAddress(sellersBsqPayout.getScriptPubKey().getToAddress(params).toString());
+            tradePeer.setPayout(sellersBsqPayout.getValue().getValue());
```

Comment:
This value is taken from the transaction received from seller but I don't see any verification of this value. Better would probably be to use the sellers bsq payout from Trade. If that's not correct the final tx won't be correct.